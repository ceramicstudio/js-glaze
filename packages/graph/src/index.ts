/**
 * Ceramic Graph client
 *
 * @module graph
 */

export { GraphClient, type GraphClientParams } from './client.js'
export { Context, type ContextParams } from './context.js'
export { createGraphQLSchema } from './schema.js'
export { printGraphQLSchema } from './utils.js'
