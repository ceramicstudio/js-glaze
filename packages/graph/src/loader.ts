import type { CeramicApi, CreateOpts, UpdateOpts } from '@ceramicnetwork/common'
import {
  ModelInstanceDocument,
  type ModelInstanceDocumentMetadata,
} from '@ceramicnetwork/stream-model-instance'
import { type CommitID, StreamID, StreamRef } from '@ceramicnetwork/streamid'
import DataLoader from 'dataloader'
import type { BatchLoadFn } from 'dataloader'

export type DocID = CommitID | StreamID | string

export type CreateOptions = CreateOpts & {
  controller?: string
}

export type UpdateDocOptions = {
  replace?: boolean
  version?: string
}

export type UpdateOptions = UpdateOpts & UpdateDocOptions

// Implements CacheMap from dataloader, copied here to generate docs
export type DocumentCache = {
  /**
   * Get a Promise of a ModelInstanceDocument by its stream ID
   */
  get(id: string): Promise<ModelInstanceDocument> | void
  /**
   * Set a Promise of a ModelInstanceDocument by its stream ID
   */
  set(id: string, value: Promise<ModelInstanceDocument>): any
  /**
   * Remove a specific entry from the cache
   */
  delete(id: string): any
  /**
   * Remove all entries from the cache
   */
  clear(): any
}

export type DocumentLoaderParams = {
  /**
   * A Ceramic client instance
   */
  ceramic: CeramicApi
  /**
   * A supported cache implementation, `true` to use the default implementation or `false` to
   * disable the cache (default)
   */
  cache?: DocumentCache | boolean
  /**
   * MultiQuery request timeout in milliseconds
   */
  multiqueryTimeout?: number
}

/** @internal */
export function idToString(id: DocID): string {
  return typeof id === 'string' ? StreamRef.from(id).toString() : id.toString()
}

const tempBatchLoadFn: BatchLoadFn<DocID, ModelInstanceDocument> = () => Promise.resolve([])

export class DocumentLoader extends DataLoader<DocID, ModelInstanceDocument> {
  #ceramic: CeramicApi
  #useCache: boolean

  constructor(params: DocumentLoaderParams) {
    super(tempBatchLoadFn, {
      cache: true, // Cache needs to be enabled for batching
      cacheKeyFn: idToString,
      cacheMap:
        params.cache != null && typeof params.cache !== 'boolean' ? params.cache : undefined,
    })

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore internal method
    this._batchLoadFn = async (keys: ReadonlyArray<DocID>) => {
      if (!params.cache) {
        // Disable cache but keep batching behavior - from https://github.com/graphql/dataloader#disabling-cache
        this.clearAll()
      }
      const results = await params.ceramic.multiQuery(
        keys.map(idToString).map((streamId) => ({ streamId })),
        params.multiqueryTimeout
      )
      return keys.map((key) => {
        const id = idToString(key)
        const doc = results[id]
        return doc
          ? (doc as unknown as ModelInstanceDocument)
          : new Error(`Failed to load document: ${id}`)
      })
    }

    this.#ceramic = params.ceramic
    this.#useCache = !!params.cache
  }

  /**
   * Add a ModelInstanceDocument to the local cache, if enabled.
   */
  cache(stream: ModelInstanceDocument): boolean {
    if (!this.#useCache) {
      return false
    }

    const id = stream.id.toString()
    this.clear(id).prime(id, stream)
    return true
  }

  /**
   * Create a new ModelInstanceDocument and add it to the cache, if enabled.
   */
  async create<T extends Record<string, any> = Record<string, any>>(
    model: string | StreamID,
    content: T,
    { controller, ...options }: CreateOptions = {}
  ): Promise<ModelInstanceDocument<T>> {
    const metadata: ModelInstanceDocumentMetadata = {
      controller: controller ?? (this.#ceramic.did?.id as string),
      model: model instanceof StreamID ? model : StreamID.fromString(model),
    }
    const stream = await ModelInstanceDocument.create<T>(this.#ceramic, content, metadata, options)
    this.cache(stream)
    return stream
  }

  /**
   * Load a ModelInstanceDocument from the cache (if enabled) or remotely.
   */
  async load<T extends Record<string, any> = Record<string, any>>(
    id: DocID
  ): Promise<ModelInstanceDocument<T>> {
    return (await super.load(id)) as ModelInstanceDocument<T>
  }

  /**
   * Update a ModelInstanceDocument after loading the stream remotely, bypassing the cache.
   */
  async update<T extends Record<string, any> = Record<string, any>>(
    streamID: string | StreamID,
    content: T,
    { replace, version, ...options }: UpdateOptions = {}
  ): Promise<ModelInstanceDocument<T>> {
    const id = idToString(streamID)
    this.clear(id)
    const stream = await this.load<T>(id)
    if (version != null && stream.commitId.toString() !== version) {
      throw new Error('Stream version mismatch')
    }
    const newContent = replace ? content : { ...stream.content, ...content }
    await stream.replace(newContent, options)
    return stream
  }
}
