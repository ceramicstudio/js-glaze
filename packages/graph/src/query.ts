import type { BaseQuery, CeramicApi, Page, Pagination, StreamState } from '@ceramicnetwork/common'
import { ModelInstanceDocument } from '@ceramicnetwork/stream-model-instance'
import type { Connection, ConnectionArguments } from 'graphql-relay'

export type ConnectionQuery = BaseQuery & ConnectionArguments
export type IndexQuery = BaseQuery & Pagination

export function toIndexQuery(source: ConnectionQuery): IndexQuery {
  const { after, before, first, last, ...base } = source
  let query: IndexQuery
  if (first != null) {
    query = { ...base, first }
    if (after != null) {
      // eslint-disable-next-line
      // @ts-ignore defined as read-only
      query.after = after
    }
    return query
  }
  if (last != null) {
    query = { ...base, last }
    if (before != null) {
      // eslint-disable-next-line
      // @ts-ignore defined as read-only
      query.before = before
    }
    return query
  }
  throw new Error('Missing "first" or "last" connection argument')
}

export function toRelayConnection(
  ceramic: CeramicApi,
  page: Page<StreamState>
): Connection<ModelInstanceDocument> {
  return {
    edges: page.edges.map(({ cursor, node }) => {
      return { cursor, node: ceramic.buildStreamFromState<ModelInstanceDocument>(node) }
    }),
    pageInfo: {
      ...page.pageInfo,
      startCursor: page.pageInfo.startCursor ?? null,
      endCursor: page.pageInfo.endCursor ?? null,
    },
  }
}

export async function queryConnection(
  ceramic: CeramicApi,
  query: ConnectionQuery
): Promise<Connection<ModelInstanceDocument>> {
  const page = await ceramic.index.queryIndex(toIndexQuery(query))
  return toRelayConnection(ceramic, page)
}

export async function querySingle(
  ceramic: CeramicApi,
  query: BaseQuery
): Promise<ModelInstanceDocument | null> {
  const result = await ceramic.index.queryIndex({ ...query, last: 1 })
  const edge = result.edges[0]
  return edge ? ceramic.buildStreamFromState<ModelInstanceDocument>(edge.node) : null
}
