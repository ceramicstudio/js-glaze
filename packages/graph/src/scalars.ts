import { StreamRef } from '@ceramicnetwork/streamid'
import { GraphQLError, GraphQLScalarType, Kind } from 'graphql'

function validateStreamRef(input: unknown): string {
  return StreamRef.from(input as string).toString()
}

export const CeramicStreamReference = new GraphQLScalarType({
  name: 'CeramicStreamReference',
  description: 'A Ceramic Stream reference (Stream or Commit ID)',
  serialize: validateStreamRef,
  parseValue: validateStreamRef,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(`Can only validate strings as StreamRef but got a: ${ast.kind}`)
    }
    return validateStreamRef(ast.value)
  },
})
