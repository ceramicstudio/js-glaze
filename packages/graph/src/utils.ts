import type { RuntimeCompositeDefinition } from '@glazed/types'
import { printSchema } from 'graphql'

import { createGraphQLSchema } from './schema.js'

/**
 * Create a GraphQL schema from a runtime composite definition and return its string
 * representation.
 */
export function printGraphQLSchema(
  definition: RuntimeCompositeDefinition,
  readonly = false
): string {
  return printSchema(createGraphQLSchema({ definition, readonly }))
}
