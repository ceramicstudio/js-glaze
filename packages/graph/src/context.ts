import type { BaseQuery, CeramicApi } from '@ceramicnetwork/common'
import type { ModelInstanceDocument } from '@ceramicnetwork/stream-model-instance'
import type { CommitID, StreamID } from '@ceramicnetwork/streamid'
import type { Connection } from 'graphql-relay'

import { type ConnectionQuery, queryConnection, querySingle } from './query.js'
import { type DocumentCache, DocumentLoader, type UpdateDocOptions } from './loader.js'

export type ContextParams = {
  /**
   * Optional cache for documents.
   */
  cache?: DocumentCache | boolean
  /**
   * Ceramic client instance.
   */
  ceramic: CeramicApi
  /**
   * @internal
   */
  loader?: DocumentLoader
}

/**
 * GraphQL execution context, exported by the {@linkcode graph} module.
 *
 * ```sh
 * import { Context } from '@glazed/graph'
 * ```
 */
export class Context {
  #ceramic: CeramicApi
  #loader: DocumentLoader

  constructor(params: ContextParams) {
    const { cache, ceramic } = params
    this.#ceramic = ceramic
    this.#loader = params.loader ?? new DocumentLoader({ ceramic, cache })
  }

  /**
   * Returns whether the Ceramic client instance used internally is authenticated or not. When not
   * authenticated, mutations will fail.
   */
  get authenticated(): boolean {
    return this.#ceramic.did?.authenticated ?? false
  }

  /**
   * Ceramic client instance used internally.
   */
  get ceramic(): CeramicApi {
    return this.#ceramic
  }

  /**
   * Document loader instance used internally.
   */
  get loader(): DocumentLoader {
    return this.#loader
  }

  /**
   * ID of the current viewer (authenticated DID), if set.
   */
  get viewerID(): string | null {
    return this.#ceramic.did?.id ?? null
  }

  /**
   * Load a document by ID, using the cache if possible.
   */
  async loadDoc<Content = Record<string, any>>(
    id: string | CommitID | StreamID,
    fresh = false
  ): Promise<ModelInstanceDocument<Content>> {
    if (fresh) {
      this.#loader.clear(id)
    }
    return await this.#loader.load<Content>(id)
  }

  /**
   * Create a new document with the given model and content.
   */
  async createDoc<Content = Record<string, any>>(
    model: string,
    content: Content
  ): Promise<ModelInstanceDocument<Content>> {
    return await this.#loader.create(model, content)
  }

  /**
   * Update an existing document.
   */
  async updateDoc<Content = Record<string, any>>(
    id: string | StreamID,
    content: Content,
    options?: UpdateDocOptions
  ): Promise<ModelInstanceDocument<Content>> {
    return await this.#loader.update(id, content, options)
  }

  /**
   * Query the index for a connection of documents.
   */
  async queryConnection(query: ConnectionQuery): Promise<Connection<ModelInstanceDocument>> {
    return await queryConnection(this.#ceramic, query)
  }

  /**
   * Query the index for a single document.
   */
  async querySingle(query: BaseQuery): Promise<ModelInstanceDocument | null> {
    return await querySingle(this.#ceramic, query)
  }
}
