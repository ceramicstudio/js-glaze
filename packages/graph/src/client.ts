import type { CeramicApi } from '@ceramicnetwork/common'
import type { RuntimeCompositeDefinition } from '@glazed/types'
import type { DID } from 'dids'
import {
  type DocumentNode,
  type ExecutionResult,
  type GraphQLError,
  type Source,
  GraphQLSchema,
  execute,
  parse,
  validate,
} from 'graphql'

import { Context } from './context.js'
import type { DocumentCache, DocumentLoader } from './loader.js'
import { createGraphQLSchema } from './schema.js'

export type GraphClientParams = {
  cache?: DocumentCache | boolean
  ceramic: CeramicApi
  definition: RuntimeCompositeDefinition
  loader?: DocumentLoader
}

export class GraphClient {
  #context: Context
  #resources: Array<string>
  #schema: GraphQLSchema

  constructor(params: GraphClientParams) {
    const { definition, ...contextParams } = params
    this.#context = new Context(contextParams)
    this.#resources = Object.values(definition.models).map((modelID) => {
      return `ceramic://*?model=${modelID}`
    })
    this.#schema = createGraphQLSchema({ definition })
  }

  get context(): Context {
    return this.#context
  }

  get did(): DID | undefined {
    return this.#context.ceramic.did
  }

  get id(): string | undefined {
    return this.did?.id
  }

  get resources(): Array<string> {
    return this.#resources
  }

  setDID(did: DID): void {
    this.#context.ceramic.did = did
  }

  async execute(
    source: string | Source,
    variableValues?: Record<string, unknown>
  ): Promise<ExecutionResult> {
    let document: DocumentNode
    try {
      document = parse(source)
    } catch (syntaxError) {
      return {
        errors: [syntaxError as GraphQLError],
      }
    }

    const validationErrors = validate(this.#schema, document)
    if (validationErrors.length > 0) {
      return {
        errors: validationErrors,
      }
    }

    return await execute({
      document,
      variableValues,
      contextValue: this.#context,
      schema: this.#schema,
    })
  }
}
