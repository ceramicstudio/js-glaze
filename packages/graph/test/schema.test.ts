import { createRuntimeDefinition } from '@glazed/devtools'

import { printGraphQLSchema } from '../src'

describe('schema', () => {
  test('printGraphQLSchema()', () => {
    const definition = createRuntimeDefinition({
      version: '1.0',
      commonEmbeds: ['ImageMetadata', 'ImageSources'],
      models: {
        genericProfileID: {
          name: 'GenericProfile',
          accountRelation: 'single',
          schema: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            type: 'object',
            properties: {
              name: {
                type: 'string',
                maxLength: 150,
              },
              image: {
                $ref: '#/definitions/imageSources',
              },
            },
            definitions: {
              IPFSURL: {
                type: 'string',
                pattern: '^ipfs://.+',
                maxLength: 150,
              },
              positiveInteger: {
                type: 'integer',
                minimum: 1,
              },
              imageMetadata: {
                type: 'object',
                title: 'ImageMetadata',
                properties: {
                  src: {
                    $ref: '#/definitions/IPFSURL',
                  },
                  mimeType: {
                    type: 'string',
                    maxLength: 50,
                  },
                  width: {
                    $ref: '#/definitions/positiveInteger',
                  },
                  height: {
                    $ref: '#/definitions/positiveInteger',
                  },
                  size: {
                    $ref: '#/definitions/positiveInteger',
                  },
                },
                required: ['src', 'mimeType', 'width', 'height'],
              },
              imageSources: {
                type: 'object',
                title: 'ImageSources',
                properties: {
                  original: {
                    $ref: '#/definitions/imageMetadata',
                  },
                  alternatives: {
                    type: 'array',
                    items: {
                      $ref: '#/definitions/imageMetadata',
                    },
                  },
                },
                required: ['original'],
              },
            },
          },
        },
        socialProfileID: {
          name: 'SocialProfile',
          accountRelation: 'single',
          schema: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            type: 'object',
            properties: {
              description: {
                type: 'string',
                maxLength: 420,
              },
              emoji: {
                type: 'string',
                maxLength: 2,
              },
              background: {
                $ref: '#/definitions/imageSources',
              },
              url: {
                type: 'string',
                maxLength: 240,
              },
            },
            definitions: {
              IPFSURL: {
                type: 'string',
                pattern: '^ipfs://.+',
                maxLength: 150,
              },
              positiveInteger: {
                type: 'integer',
                minimum: 1,
              },
              imageMetadata: {
                type: 'object',
                title: 'ImageMetadata',
                properties: {
                  src: {
                    $ref: '#/definitions/IPFSURL',
                  },
                  mimeType: {
                    type: 'string',
                    maxLength: 50,
                  },
                  width: {
                    $ref: '#/definitions/positiveInteger',
                  },
                  height: {
                    $ref: '#/definitions/positiveInteger',
                  },
                  size: {
                    $ref: '#/definitions/positiveInteger',
                  },
                },
                required: ['src', 'mimeType', 'width', 'height'],
              },
              imageSources: {
                type: 'object',
                title: 'ImageSources',
                properties: {
                  original: {
                    $ref: '#/definitions/imageMetadata',
                  },
                  alternatives: {
                    type: 'array',
                    items: {
                      $ref: '#/definitions/imageMetadata',
                    },
                  },
                },
                required: ['original'],
              },
            },
          },
        },
        personProfileID: {
          name: 'PersonProfile',
          accountRelation: 'single',
          schema: {
            $schema: 'http://json-schema.org/draft-07/schema#',
            type: 'object',
            properties: {
              birthDate: {
                type: 'string',
                format: 'date',
                maxLength: 10,
              },
              gender: {
                type: 'string',
                maxLength: 42,
              },
              homeLocation: {
                type: 'string',
                maxLength: 140,
              },
              residenceCountry: {
                type: 'string',
                pattern: '^[A-Z]{2}$',
                maxLength: 2,
              },
              nationalities: {
                type: 'array',
                minItems: 1,
                items: {
                  type: 'string',
                  pattern: '^[A-Z]{2}$',
                  maxItems: 5,
                },
              },
              affiliations: {
                type: 'array',
                items: {
                  type: 'string',
                  maxLength: 140,
                },
              },
            },
          },
        },
      },
    })
    expect(printGraphQLSchema(definition)).toMatchSnapshot()
  })
})
