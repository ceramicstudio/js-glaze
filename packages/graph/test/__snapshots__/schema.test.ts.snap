// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema printGraphQLSchema() 1`] = `
"type Query {
  \\"\\"\\"Fetches an object given its ID\\"\\"\\"
  node(
    \\"\\"\\"The ID of an object\\"\\"\\"
    id: ID!
  ): Node
  account(id: DID!): CeramicAccount!
  viewer: CeramicAccount
}

\\"\\"\\"An object with an ID\\"\\"\\"
interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}

type CeramicAccount {
  id: DID!
  isViewer: Boolean!
  data: CeramicAccountData!
}

\\"\\"\\"
A field whose value conforms to the standard DID format as specified in did-core: https://www.w3.org/TR/did-core/.
\\"\\"\\"
scalar DID @specifiedBy(url: \\"https://www.w3.org/TR/did-core/\\")

type CeramicAccountData {
  genericProfile: GenericProfile
  socialProfile: SocialProfile
  personProfile: PersonProfile
}

type GenericProfile implements Node {
  id: ID!
  name: String
  image: ImageSources
}

type ImageSources {
  original: ImageMetadata!
  alternatives: [ImageMetadata]
}

type ImageMetadata {
  src: String!
  mimeType: String!
  width: Int!
  height: Int!
  size: Int
}

type SocialProfile implements Node {
  id: ID!
  description: String
  emoji: String
  background: ImageSources
  url: String
}

type PersonProfile implements Node {
  id: ID!
  birthDate: String
  gender: String
  homeLocation: String
  residenceCountry: String
  nationalities: [String]
  affiliations: [String]
}

type Mutation {
  createGenericProfile(input: CreateGenericProfileInput!): CreateGenericProfilePayload
  updateGenericProfile(input: UpdateGenericProfileInput!): UpdateGenericProfilePayload
  createSocialProfile(input: CreateSocialProfileInput!): CreateSocialProfilePayload
  updateSocialProfile(input: UpdateSocialProfileInput!): UpdateSocialProfilePayload
  createPersonProfile(input: CreatePersonProfileInput!): CreatePersonProfilePayload
  updatePersonProfile(input: UpdatePersonProfileInput!): UpdatePersonProfilePayload
}

type CreateGenericProfilePayload {
  node: GenericProfile!
  clientMutationId: String
}

input CreateGenericProfileInput {
  content: RequiredGenericProfileInput!
  clientMutationId: String
}

input RequiredGenericProfileInput {
  name: String
  image: RequiredImageSourcesInput
}

input RequiredImageSourcesInput {
  original: RequiredImageMetadataInput!
  alternatives: [RequiredImageMetadataInput]
}

input RequiredImageMetadataInput {
  src: String!
  mimeType: String!
  width: Int!
  height: Int!
  size: Int
}

type UpdateGenericProfilePayload {
  node: GenericProfile!
  clientMutationId: String
}

input UpdateGenericProfileInput {
  id: ID!
  content: GenericProfileInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input GenericProfileInput {
  name: String
  image: ImageSourcesInput
}

input ImageSourcesInput {
  original: ImageMetadataInput
  alternatives: [ImageMetadataInput]
}

input ImageMetadataInput {
  src: String
  mimeType: String
  width: Int
  height: Int
  size: Int
}

input UpdateOptionsInput {
  replace: Boolean
  version: CeramicCommitID
}

\\"\\"\\"A Ceramic Commit ID\\"\\"\\"
scalar CeramicCommitID

type CreateSocialProfilePayload {
  node: SocialProfile!
  clientMutationId: String
}

input CreateSocialProfileInput {
  content: RequiredSocialProfileInput!
  clientMutationId: String
}

input RequiredSocialProfileInput {
  description: String
  emoji: String
  background: RequiredImageSourcesInput
  url: String
}

type UpdateSocialProfilePayload {
  node: SocialProfile!
  clientMutationId: String
}

input UpdateSocialProfileInput {
  id: ID!
  content: SocialProfileInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input SocialProfileInput {
  description: String
  emoji: String
  background: ImageSourcesInput
  url: String
}

type CreatePersonProfilePayload {
  node: PersonProfile!
  clientMutationId: String
}

input CreatePersonProfileInput {
  content: RequiredPersonProfileInput!
  clientMutationId: String
}

input RequiredPersonProfileInput {
  birthDate: String
  gender: String
  homeLocation: String
  residenceCountry: String
  nationalities: [String]
  affiliations: [String]
}

type UpdatePersonProfilePayload {
  node: PersonProfile!
  clientMutationId: String
}

input UpdatePersonProfileInput {
  id: ID!
  content: PersonProfileInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input PersonProfileInput {
  birthDate: String
  gender: String
  homeLocation: String
  residenceCountry: String
  nationalities: [String]
  affiliations: [String]
}"
`;

exports[`schema printGraphQLSchema() 2`] = `
"type Query {
  \\"\\"\\"Fetches an object given its ID\\"\\"\\"
  node(
    \\"\\"\\"The ID of an object\\"\\"\\"
    id: ID!
  ): Node
  account(id: DID!): CeramicAccount!
  viewer: CeramicAccount
}

\\"\\"\\"An object with an ID\\"\\"\\"
interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}

type CeramicAccount {
  id: DID!
  isViewer: Boolean!
  data: CeramicAccountData!
}

\\"\\"\\"
A field whose value conforms to the standard DID format as specified in did-core: https://www.w3.org/TR/did-core/.
\\"\\"\\"
scalar DID @specifiedBy(url: \\"https://www.w3.org/TR/did-core/\\")

type CeramicAccountData {
  noteCollection(
    \\"\\"\\"Returns the items in the list that come after the specified cursor.\\"\\"\\"
    after: String

    \\"\\"\\"Returns the first n items from the list.\\"\\"\\"
    first: Int

    \\"\\"\\"Returns the items in the list that come before the specified cursor.\\"\\"\\"
    before: String

    \\"\\"\\"Returns the last n items from the list.\\"\\"\\"
    last: Int
  ): NoteConnection
}

\\"\\"\\"A connection to a list of items.\\"\\"\\"
type NoteConnection {
  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"A list of edges.\\"\\"\\"
  edges: [NoteEdge]
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: String

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: String
}

\\"\\"\\"An edge in a connection.\\"\\"\\"
type NoteEdge {
  \\"\\"\\"The item at the end of the edge\\"\\"\\"
  node: Note

  \\"\\"\\"A cursor for use in pagination\\"\\"\\"
  cursor: String!
}

type Note implements Node {
  id: ID!
  title: String!
  text: String!
  author: CeramicAccount!
  version: CeramicCommitID!
}

\\"\\"\\"A Ceramic Commit ID\\"\\"\\"
scalar CeramicCommitID

type Mutation {
  createNote(input: CreateNoteInput!): CreateNotePayload
  updateNote(input: UpdateNoteInput!): UpdateNotePayload
}

type CreateNotePayload {
  node: Note!
  clientMutationId: String
}

input CreateNoteInput {
  content: RequiredNoteInput!
  clientMutationId: String
}

input RequiredNoteInput {
  title: String!
  text: String!
}

type UpdateNotePayload {
  node: Note!
  clientMutationId: String
}

input UpdateNoteInput {
  id: ID!
  content: NoteInput!
  options: UpdateOptionsInput
  clientMutationId: String
}

input NoteInput {
  title: String
  text: String
}

input UpdateOptionsInput {
  replace: Boolean
  version: CeramicCommitID
}"
`;
